db/system_keyspace.cc

```diff
@@ -270,7 +271,7 @@ schema_ptr system_keyspace::topology_requests() {
             .with_column("error", utf8_type)
             .with_column("end_time", timestamp_type)
             .set_comment("Topology request tracking")
-            .with_version(generate_schema_version(id))
+            .with_version(generate_schema_version(id, schema_version_offset))
             .build();
```

Comment:
How does that work during upgrades? Some nodes have the new schema, some the old?

Previously we used cluster features and schema_features to incrementally upgrade the schema.

/cc @kbr-scylla

Reply:
The cluster feature was used for that column originally - it is used to ensure that no data is written to `request_type` column. Schema features weren't - they are used for `schema_tables` only I think.

In the issue, I was told to look at the `redact_columns_for_missing_features`. The function upgrades mutations not to contain the new column during upgrade. I tried something analogical for `request_type` column and it didn't work - unsurprisingly, no data was added to the new column so mutations didn't contain it.

After that I recalled the discussion about the new column order, so I moved the new column to the end, but it still didn't work.

I went through the code and the old related changes and finally I encountered 2f49eeb

Reply:
Where works = bug reproducer from issue passes

Reply:
@tgrabiec explained this to me like 3 times in the past but I always forget

Reply:
@Deexie did you decode the backtrace from the issue? What is it?

I guess what could happen is that on the sender of the snapshot (node with new version) serializes the mutation and even though the mutation does not contain the new column, **some other columns get assigned a greater `column_id`** (perhaps the IDs of columns are not defined by how they are listed in the builder, but by some other, arbitrary way -- so when you add the new column, it ends up with some ID in the middle, which shifts the IDs of other columns up). So when the receiver tries to deserialize the `canonical_mutation`, it sees that some column has ID `5` (when it would be `4` if the mutation was constructed locally -- generally there is a mismatch between IDs).

But I don't understand why assigning a different schema version fixes the problem.

@tgrabiec do you remember why it helps?

Reply:
```cpp
mutation canonical_mutation::to_mutation(schema_ptr s) const {
    auto in = ser::as_input_stream(_data);
    auto mv = ser::deserialize(in, boost::type<ser::canonical_mutation_view>());

    auto cf_id = mv.table_id();
    if (s->id() != cf_id) {
        throw std::runtime_error(format("Attempted to deserialize canonical_mutation of table {} with schema of table {} ({}.{})",
                                        cf_id, s->id(), s->ks_name(), s->cf_name()));
    }

    auto version = mv.schema_version();
    auto pk = mv.key();

    mutation m(std::move(s), std::move(pk));

    if (version == m.schema()->version()) {
        auto partition_view = mutation_partition_view::from_view(mv.partition());
        mutation_application_stats app_stats;
        m.partition().apply(*m.schema(), partition_view, *m.schema(), app_stats);
    } else {
        column_mapping cm = mv.mapping();
        converting_mutation_partition_applier v(cm, *m.schema(), m.partition());
        auto partition_view = mutation_partition_view::from_view(mv.partition());
        partition_view.accept(cm, v);
    }
    return m;
}

```

Ok I'm guessing that we're taking the wrong branch here if versions are equal (when they shouldn't be)

Reply:
That must indeed be it
to translate `canonical_mutation` to `mutation`, we use `canonical_mutation::to_mutation` and pass `schema_ptr` **obtained locally based on the `column_family_id()`**:
```cpp
        for (const auto& cm : canonical_mutations) {
            auto& tbl = db.find_column_family(cm.column_family_id());
            mutations.emplace_back(cm.to_mutation(
                    tbl.schema()));
```

the `colum_family_id()` is the same on snapshot puller and snapshot sender. So we find the schema locally.

Of course the local `schema` object does not contain the column, the remote one does.

Now if we did not change the version, we take this branch
```cpp
    if (version == m.schema()->version()) {
```
so column mapping is taken from local schema version, and we see the `on_internal_error`.

If we did change the version then we take the `else` branch and mapping is taken from `canonical_mutation` (it stores the mapping)
```cpp
        column_mapping cm = mv.mapping();
        converting_mutation_partition_applier v(cm, *m.schema(), m.partition());
```

So one thing I don't understand --- why do we need the first branch at all? Wouldn't it be correct to always take column mapping from `canonical_mutation`? Maybe then we wouldn't have to remember to update the version.

@tgrabiec do you know?

Reply:
@tgrabiec explained to me on slack -- the first branch is an optimization; going through `converting_mutation_partition_applier` is costly